src/stream/src/executor/arrangement_backfill.rs

```diff
@@ -0,0 +1,667 @@
+        // scan. Otherwise, use Excluded.
+        let range_bounds: (Bound<OwnedRow>, Bound<OwnedRow>) =
+            if let Some(current_pos) = current_pos {
+                // If `current_pos` is an empty row which means upstream mv contains only one row
+                // and it has been consumed. The iter interface doesn't support
+                // `Excluded(empty_row)` range bound, so we can simply return `None`.
+                if current_pos.is_empty() {
+                    assert!(upstream_table.pk_indices().is_empty());
+                    yield None;
+                    return Ok(());
+                }
+
+                (Bound::Excluded(current_pos), Bound::Unbounded)
+            } else {
+                (Bound::Unbounded, Bound::Unbounded)
+            };
+        let iterators = upstream_table
+            .iter_all_vnodes_with_pk_range(&range_bounds, Default::default())
+            .await?;
+        let pinned_iter: Vec<_> = iterators.into_iter().map(Box::pin).collect_vec();
+        let iter = merge_sort(pinned_iter);
+        pin_mut!(iter);
+        while let Some(data_chunk) = collect_data_chunk(&mut iter, &schema, Some(CHUNK_SIZE))
+            .instrument_await("arrangement_backfill_snapshot_read")
+            .await?
+        {
+            if data_chunk.cardinality() != 0 {
+                let ops = vec![Op::Insert; data_chunk.capacity()];
+                let stream_chunk = StreamChunk::from_parts(ops, data_chunk);
+                yield Some(stream_chunk);
+            }
+        }
+
+        yield None;
+    }
+
+    /// Mark chunk:
+    /// For each row of the chunk, forward it to downstream if its pk <= `current_pos`, otherwise
+    /// ignore it. We implement it by changing the visibility bitmap.
+    fn mark_chunk(
+        chunk: &StreamChunk,
+        current_pos: &OwnedRow,
+        pk_in_output_indices: PkIndicesRef<'_>,
+        pk_order: &[OrderType],
+    ) -> StreamChunk {
+        // FIXME: Temporary workaround, we can optimize it later when benchmarking.
+        // We can do so with a `chunk.compact_ref()` instead of `chunk.compact()`.
+        let chunk = chunk.clone();
+        let chunk = chunk.compact();
+        let (data, ops) = chunk.into_parts();
+        let mut new_visibility = BitmapBuilder::with_capacity(ops.len());
+        // Use project to avoid allocation.
+        for v in data.rows().map(|row| {
+            match row
+                .project(pk_in_output_indices)
+                .iter()
+                .zip_eq_fast(pk_order.iter().copied())
+                .cmp_by(current_pos.iter(), |(x, order), y| cmp_datum(x, y, order))
+            {
+                Ordering::Less | Ordering::Equal => true,
+                Ordering::Greater => false,
+            }
+        }) {
+            new_visibility.append(v);
+        }
+        let (columns, _) = data.into_parts();
+        StreamChunk::new(ops, columns, Some(new_visibility.finish()))
+    }
+
```

Comment:
Everything after this point should be the same as `backfill_executor`, (except for `state_table`, code handling backwards compat has been removed).